src/operator/numpy/np_matmul_op-inl.h

```diff
@@ -0,0 +1,387 @@
+    MatmulImpl<xpu, DType>(ctx, inputs[0], inputs[1], req[0], outputs[0], temp_mem,
+                           ndim, batch_size, bc_size_a, bc_size_b,
+                           a_shape, b_shape, out_shape, false, false);
+  });
+}
+
+
+template<typename xpu>
+void NumpyMatmulBackward(const nnvm::NodeAttrs& attrs,
+                        const OpContext& ctx,
+                        const std::vector<TBlob>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  if (req[0] == kNullOp) return;
+
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 2U);
+  
+  const TBlob& ograd = inputs[0];
+  const TBlob& a = inputs[1];
+  const TBlob& b = inputs[2];
+  const TBlob& grad_a = outputs[0];
+  const TBlob& grad_b = outputs[1];
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+
+  CHECK(grad_a.type_flag_ == kFloat32 || grad_a.type_flag_ == kFloat64 ||
+    (grad_a.type_flag_ == kFloat16 && ctx.run_ctx.ctx.dev_mask() == mshadow::gpu::kDevMask))
+    << "Matmul only supports float32/float64 for CPU, and float16/float32/float64 for GPU";
+  CHECK(grad_b.type_flag_ == kFloat32 || grad_b.type_flag_ == kFloat64 ||
+    (grad_b.type_flag_ == kFloat16 && ctx.run_ctx.ctx.dev_mask() == mshadow::gpu::kDevMask))
+    << "Matmul only supports float32/float64 for CPU, and float16/float32/float64 for GPU";
+  if ((a.shape_.Size() == 0) || (b.shape_.Size() == 0 || (ograd.shape_.Size() == 0))) {
+    MSHADOW_REAL_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      Kernel<mxnet_op::set_zero, xpu>::Launch(s, grad_a.shape_.Size(), grad_a.dptr<DType>());
+      Kernel<mxnet_op::set_zero, xpu>::Launch(s, grad_b.shape_.Size(), grad_b.dptr<DType>());
+    });
+    return;
+  }
+
+  mxnet::TShape a_shape = (a.shape_.ndim() == 1) ? Shape2(1, a.shape_.Size()) : a.shape_;
+  mxnet::TShape b_shape = (b.shape_.ndim() == 1) ? Shape2(b.shape_.Size(), 1) : b.shape_;
+  mxnet::TShape out_shape = ograd.shape_;
+  size_t ndim = out_shape.ndim();
+  if ((a.shape_.ndim() == 1) && (b.shape_.ndim() == 1)) {
+    ndim = 2;
+    std::vector<size_t> newshape({1, 1});
+    out_shape.assign(newshape.begin(), newshape.end());
+  } else if ((a.shape_.ndim() == 1) && (b.shape_.ndim() != 1)) {
+    ndim = out_shape.ndim() + 1;
+    std::vector<size_t> newshape(ndim);
+    for (size_t i = 0; i < ndim - 2; ++i) {
+      newshape[i] = out_shape[i];
+    }
+    newshape[ndim - 2] = 1;
+    newshape[ndim - 1] = out_shape[ndim - 2];
+    out_shape.assign(newshape.begin(), newshape.end());
+  } else if ((a.shape_.ndim() != 1) && (b.shape_.ndim() == 1)) {
+    ndim = out_shape.ndim() + 1;
+    std::vector<size_t> newshape(ndim);
+    for (size_t i = 0; i < ndim - 1; ++i) {
+      newshape[i] = out_shape[i];
+    }
+    newshape[ndim - 1] = 1;
+    out_shape.assign(newshape.begin(), newshape.end());
```

Comment:
seems like `out_shape.assign(newshape.begin(), newshape.end());` is shared by all branches, maybe you could put it outside.

Reply:
A more general case is that  a.ndim !=1 && b.ndim !=1, which don't need to assign 'out_shape' with 'newshape'. If I put this operation outside, I need to fill 'newshape' with original 'out_shape' and then fill final 'out_shape' with 'newshape' for that case, which need redundant process.